// @phonepe-pg/pg-sdk-node/dist/payments/v2/models/request/CustomCheckoutPayRequest.js
"use strict";
/*
 *  Copyright (c) 2025 Original Author(s), PhonePe India Pvt. Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomCheckoutPayRequest = void 0;
const UpiIntentPayRequestBuilder_1 = require("../../../../common/models/request/instruments/UpiIntentPayRequestBuilder");
const UpiCollectPayViaVpaRequestBuilder_1 = require("../../../../common/models/request/instruments/UpiCollectPayViaVpaRequestBuilder");
const UpiQrRequestBuilder_1 = require("../../../../common/models/request/instruments/UpiQrRequestBuilder");
const NetBankingPayRequestBuilder_1 = require("../../../../common/models/request/instruments/NetBankingPayRequestBuilder");
const TokenPayRequestBuilder_1 = require("../../../../common/models/request/instruments/TokenPayRequestBuilder");
const CardPayRequestBuilder_1 = require("../../../../common/models/request/instruments/CardPayRequestBuilder");
/**
 * Creates a request based on different builder for different instruments.
 * 1. UpiIntentPayRequestBuilder -> Upi Intent
 * 2. UpiCollectPayViaVpaRequestBuilder -> Upi Collect
 * 3. UpiQrRequestBuilder -> Upi QR
 * 4. NetBankingPayRequestBuilder -> Netbanking
 * 5. TokenPayRequestBuilder -> Token
 * 6. CardPayRequestBuilder -> Card
 */
class CustomCheckoutPayRequest {
    constructor(merchantOrderId, amount, paymentFlow, expireAfter, metaInfo, constrainsts, deviceContext, deviceOS) {
        this.merchantOrderId = merchantOrderId;
        this.amount = amount;
        this.paymentFlow = paymentFlow;
        this.metaInfo = metaInfo;
        this.constraints = constrainsts;
        this.deviceContext = deviceContext;
        this.expireAfter = expireAfter;
        this.deviceOS = deviceOS;
    }
    toJSON() {
        const _a = this, { deviceOS } = _a, rest = __rest(_a, ["deviceOS"]);
        return rest;
    }
    static UpiIntentPayRequestBuilder() {
        return new UpiIntentPayRequestBuilder_1.UpiIntentPayRequestBuilder();
    }
    static UpiCollectPayViaVpaRequestBuilder() {
        return new UpiCollectPayViaVpaRequestBuilder_1.UpiCollectPayViaVpaRequestBuilder();
    }
    static UpiQrRequestBuilder() {
        return new UpiQrRequestBuilder_1.UpiQrRequestBuilder();
    }
    static NetBankingPayRequestBuilder() {
        return new NetBankingPayRequestBuilder_1.NetBankingPayRequestBuilder();
    }
    static TokenPayRequestBuilder() {
        return new TokenPayRequestBuilder_1.TokenPayRequestBuilder();
    }
    static CardPayRequestBuilder() {
        return new CardPayRequestBuilder_1.CardPayRequestBuilder();
    }
}
exports.CustomCheckoutPayRequest = CustomCheckoutPayRequest;
