// @phonepe-pg/pg-sdk-node/dist/common/models/MetaInfo.js
"use strict";
/*
 *  Copyright (c) 2025 Original Author(s), PhonePe India Pvt. Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetaInfo = void 0;
class MetaInfo {
    constructor(udf1, udf2, udf3, udf4, udf5, udf6, udf7, udf8, udf9, udf10, udf11, udf12, udf13, udf14, udf15) {
        const freeFields = [udf1, udf2, udf3, udf4, udf5, udf6, udf7, udf8, udf9, udf10];
        freeFields.forEach((val, idx) => MetaInfo._validateSize(`udf${idx + 1}`, val, MetaInfo.FREE_MAX));
        const restrictedFields = [udf11, udf12, udf13, udf14, udf15];
        restrictedFields.forEach((val, idx) => MetaInfo._validateSizeAndPattern(`udf${idx + 11}`, val, MetaInfo.RESTRICTED_MAX));
        this.udf1 = udf1;
        this.udf2 = udf2;
        this.udf3 = udf3;
        this.udf4 = udf4;
        this.udf5 = udf5;
        this.udf6 = udf6;
        this.udf7 = udf7;
        this.udf8 = udf8;
        this.udf9 = udf9;
        this.udf10 = udf10;
        this.udf11 = udf11;
        this.udf12 = udf12;
        this.udf13 = udf13;
        this.udf14 = udf14;
        this.udf15 = udf15;
    }
    static _validateSize(field, value, max) {
        if (value !== undefined && value.length > max) {
            throw new Error(`${field} exceeds maximum allowed size of ${max} characters`);
        }
    }
    static _validateSizeAndPattern(field, value, max) {
        if (value === undefined)
            return;
        if (value.length > max) {
            throw new Error(`${field} exceeds maximum allowed size of ${max} characters`);
        }
        if (!MetaInfo.RESTRICTED_PATTERN.test(value)) {
            throw new Error(`${field} should only contain alphanumeric characters, underscores, hyphens, spaces, @, ., and +`);
        }
    }
    static builder() {
        return new MetaInfoBuilder();
    }
}
exports.MetaInfo = MetaInfo;
MetaInfo.FREE_MAX = 256;
MetaInfo.RESTRICTED_MAX = 50;
MetaInfo.RESTRICTED_PATTERN = /^[a-zA-Z0-9_\- @.+]*$/;
class MetaInfoBuilder {
    constructor() {
        /**
         * SETTERS
         */
        this.udf1 = (udf1) => {
            this._udf1 = udf1;
            return this;
        };
        this.udf2 = (udf2) => {
            this._udf2 = udf2;
            return this;
        };
        this.udf3 = (udf3) => {
            this._udf3 = udf3;
            return this;
        };
        this.udf4 = (udf4) => {
            this._udf4 = udf4;
            return this;
        };
        this.udf5 = (udf5) => {
            this._udf5 = udf5;
            return this;
        };
        this.udf6 = (udf6) => {
            this._udf6 = udf6;
            return this;
        };
        this.udf7 = (udf7) => {
            this._udf7 = udf7;
            return this;
        };
        this.udf8 = (udf8) => {
            this._udf8 = udf8;
            return this;
        };
        this.udf9 = (udf9) => {
            this._udf9 = udf9;
            return this;
        };
        this.udf10 = (udf10) => {
            this._udf10 = udf10;
            return this;
        };
        this.udf11 = (udf11) => {
            this._udf11 = udf11;
            return this;
        };
        this.udf12 = (udf12) => {
            this._udf12 = udf12;
            return this;
        };
        this.udf13 = (udf13) => {
            this._udf13 = udf13;
            return this;
        };
        this.udf14 = (udf14) => {
            this._udf14 = udf14;
            return this;
        };
        this.udf15 = (udf15) => {
            this._udf15 = udf15;
            return this;
        };
        this.build = () => {
            return new MetaInfo(this._udf1, this._udf2, this._udf3, this._udf4, this._udf5, this._udf6, this._udf7, this._udf8, this._udf9, this._udf10, this._udf11, this._udf12, this._udf13, this._udf14, this._udf15);
        };
    }
}
